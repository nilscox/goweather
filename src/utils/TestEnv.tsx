import * as React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';

import store from '../store';

type TestEnvProps = {
  children: React.ReactNode;
};

const TestEnv = ({ children }: TestEnvProps) => (
  <Provider store={store}>
    <Router>
      { children }
    </Router>
  </Provider>
);

export default TestEnv;
