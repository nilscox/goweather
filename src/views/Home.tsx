/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import { Component } from 'react';
import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';
import { Button, Container, Input, InputGroup } from 'reactstrap';

import { PageTitle } from '../components';
import { ICity } from '../interfaces';
import { fetchWeatherFromCityName } from '../store/actions';
import { getHistory, addToHistory } from '../services/history-service';

type HomeProps = {
  fetchWeather: (cityName: string, countryCode: string) => any; // code smell
};

type HomeState = {
  cityName: string;
  countryCode: string;
  redirectCityId: number | null;
};

const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchWeather: (cityName: string, countryCode: string) => {
    return dispatch(fetchWeatherFromCityName(cityName, countryCode));
  },
});

/**
 * Home page. Displays a form allowing to search for a city, along with some information.
 * When a city is searched (and found by the API), it is then added to the search history.
 */
class Home extends Component<HomeProps, HomeState> {

  public state = {
    cityName: '',
    countryCode: '',
    redirectCityId: null,
  };

  public componentDidMount() {
    const history: ICity[] = getHistory();

    if (history.length > 0) {
      this.setState({
        cityName: history[0].name,
        countryCode: history[0].country,
      });
    }
  }

  public render() {
    const { redirectCityId } = this.state;

    if (redirectCityId)
      return <Redirect to={'/forecast/' + redirectCityId} />;

    return (
      <Container className="pb-4">

        <PageTitle>Welcome to GoWeather!</PageTitle>

        <p>
          Search for a location using the form below to see its weather forecast for the 5 next days.
        </p>

        { this.renderForm() }

        <p>
          This website allows you to quickly access the weather forecast for any city.
          You can also view the last search queries using the <em><Link to="/history">History</Link></em> link
          in the toolbar, to quickly retrieve the cities you frequently search for.
        </p>

        <p>
          As you also may have noticed, there is a "Log in" button in the toolbar too. This feature is
          currently under development, and you will soon be informed of its avaibility if you register
          to the newsletter! But as the newsletter is also under development, don't forget to come around
          and see our progress on this website directly...
        </p>

        <p>
          We hope you will find here the information you need. Feel free to contact us to submit a bug
          report, or fill a feature request!
        </p>

      </Container>
    );
  }

  private renderForm() {
    const { cityName, countryCode } = this.state;

    return (
      <form css={formStyle} className="py-3 px-2" onSubmit={e => this.handleSubmit(e)}>

        <InputGroup className="my-1">
          <Input
            placeholder="City"
            value={cityName}
            onChange={e => this.setState({ cityName: e.currentTarget.value })}
          />
        </InputGroup>

        <InputGroup className="my-1">
          <Input
            placeholder="Country code"
            value={countryCode}
            onChange={e => this.setState({ countryCode: e.currentTarget.value })}
          />
        </InputGroup>

        <Button type="submit">Submit</Button>

      </form>
    );
  }

  private async handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();

    const cityName = this.state.cityName.trim();
    const countryCode = this.state.countryCode.trim();

    if (!cityName.length || !countryCode.length)
      return;

    const { payload } = await this.props.fetchWeather(cityName, countryCode);
    const { res, json } = payload;

    if (res.ok) {
      const { city } = json;

      addToHistory({ id: city.id, name: city.name, country: city.country });
      this.setState({ redirectCityId: city.id });
    }
  }

}

const formStyle = css`
  max-width: 450px;
`;

export default connect(null, mapDispatchToProps)(Home);
