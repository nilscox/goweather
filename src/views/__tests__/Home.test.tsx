import * as React from 'react';
import * as renderer from 'react-test-renderer';

import TestEnv from '../../utils/TestEnv';

import Home from '../Home';

test('Home', () => {
  const tree = renderer
    .create(<TestEnv><Home /></TestEnv>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
