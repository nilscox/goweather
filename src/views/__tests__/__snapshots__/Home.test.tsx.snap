// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home 1`] = `
<div
  className="pb-4 container"
>
  <h1
    className="text-center py-4 css-1aylpg"
  >
    Welcome to GoWeather!
  </h1>
  <p>
    Search for a location using the form below to see its weather forecast for the 5 next days.
  </p>
  <form
    className="py-3 px-2 css-1bivl2p"
    onSubmit={[Function]}
  >
    <div
      className="my-1 input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="City"
        type="text"
        value=""
      />
    </div>
    <div
      className="my-1 input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Country code"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label={null}
      className="btn btn-secondary"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
  <p>
    This website allows you to quickly access the weather forecast for any city. You can also view the last search queries using the 
    <em>
      <a
        href="/history"
        onClick={[Function]}
      >
        History
      </a>
    </em>
     link in the toolbar, to quickly retrieve the cities you frequently search for.
  </p>
  <p>
    As you also may have noticed, there is a "Log in" button in the toolbar too. This feature is currently under development, and you will soon be informed of its avaibility if you register to the newsletter! But as the newsletter is also under development, don't forget to come around and see our progress on this website directly...
  </p>
  <p>
    We hope you will find here the information you need. Feel free to contact us to submit a bug report, or fill a feature request!
  </p>
</div>
`;
